staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go

```diff
@@ -284,27 +282,32 @@ func (s *SpdyRoundTripper) tlsConn(ctx context.Context, rwc net.Conn, targetHost
 // dialWithoutProxy dials the host specified by url, using TLS if appropriate.
 func (s *SpdyRoundTripper) dialWithoutProxy(ctx context.Context, url *url.URL) (net.Conn, error) {
 	dialAddr := netutil.CanonicalAddr(url)
+	netD := s.Dialer
+	if netD == nil {
+		netD = &net.Dialer{
+			Timeout: 30 * time.Second,
+		}
+	}
 
 	if url.Scheme == "http" {
-		if s.Dialer == nil {
-			var d net.Dialer
-			return d.DialContext(ctx, "tcp", dialAddr)
-		} else {
-			return s.Dialer.DialContext(ctx, "tcp", dialAddr)
-		}
+		return netD.DialContext(ctx, "tcp", dialAddr)
 	}
 
-	// TODO validate the TLSClientConfig is set up?
-	var conn *tls.Conn
-	var err error
-	if s.Dialer == nil {
-		conn, err = tls.Dial("tcp", dialAddr, s.tlsConfig)
-	} else {
-		conn, err = tls.DialWithDialer(s.Dialer, "tcp", dialAddr, s.tlsConfig)
+	tlsD := tls.Dialer{
+		NetDialer: netD,
+		Config:    s.tlsConfig,
 	}
+	success := false
+	c, err := tlsD.DialContext(ctx, "tcp", dialAddr)
 	if err != nil {
 		return nil, err
 	}
+	defer func() {
+		if !success {
+			c.Close()
```

Comment:
The connection is now always closed on error. It was leaked before (on error).

Reply:
I'm not sure is leaked, it looks to me that it will be closed later

```
// Dial implements k8s.io/apimachinery/pkg/util/net.Dialer.
func (s *SpdyRoundTripper) Dial(req *http.Request) (net.Conn, error) {
	conn, err := s.dial(req)
	if err != nil {
		return nil, err
	}

	if err := req.Write(conn); err != nil {
		conn.Close()
		return nil, err
	}

	return conn, nil
}
```

Reply:
hmm, I still think the conn will not leak

```
func (s *SpdyRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
	header := utilnet.CloneHeader(req.Header)
	header.Add(httpstream.HeaderConnection, httpstream.HeaderUpgrade)
	header.Add(httpstream.HeaderUpgrade, HeaderSpdy31)

	var (
		conn        net.Conn
		rawResponse []byte
		err         error
	)

	clone := utilnet.CloneRequest(req)
	clone.Header = header
	conn, err = s.Dial(clone)
	if err != nil {
		return nil, err
	}
```

is not that it runs in a goroutine, once the RoundTrip exits it will be destroyed, no?

Reply:
A connection is a kind of object that needs to get it's lifetime managed by the program. GC will collect the object, but I don't know if there is a finalizer to clean up the unreleased underlying OS resources. It's poor practice to rely on finalizers anyway. I don't understand what is your concern. Why not close the connection if it's not needed (because an error is returned)?

Reply:
because this code is too old that there are a lot of assumptions that we don't know and we may break something without noticing, I'm sure you are right, but what I'm not sure is about the consequences of the changes, that is why I'm trying to be conservative ... is not the first time that for fixing something a bug arise in other place of the stack, and then is very hard to solve these bugs ... we can open a bug for the conns leaks and follow up later, I think that is not blocking your problem, and I can work on the test if you want